### ISSUE TITLE
Adding trainable variable to PDEs

### ISSUE THREAD

**Issue by LordPontiag** [lululxvi/deepxde#1956]
Dear @lululxvi , @praksharma
I hope this message finds you well,
I'm writing to ask for help.

I'm solving the Navier-Stokes (NS) equations using DeepXDE with a TensorFlow backend. I have a dataset of solutions (DNS data), and I'm using PointCloud for geometry and PointSetBC along with anchors to set the data. Currently, the results are similar to the DNS data I have. Now, I want to modify the NS coefficients during the training and testing process. For instance, here are my NS PDEs:

    def pde(X,Y):

      du_x =dde.grad.jacobian(Y, X, i=0, j=0)
      du_y =dde.grad.jacobian(Y, X, i=0, j=1)
      dv_x =dde.grad.jacobian(Y, X, i=1, j=0)
      dv_y =dde.grad.jacobian(Y, X, i=1, j=1)
      dp_x =dde.grad.jacobian(Y, X, i=2, j=0)
      dp_y =dde.grad.jacobian(Y, X, i=2, j=1)
  
      du_xx = dde.grad.hessian(Y, X, component=0, i=0, j=0)
      du_yy = dde.grad.hessian(Y, X, component=0, i=1, j=1)
      dv_xx = dde.grad.hessian(Y, X, component=1, i=0, j=0)
      dv_yy = dde.grad.hessian(Y, X, component=1, i=1, j=1)
      
      pde_u    = Y[:,0:1]*du_x +  Y[:,1:2]*du_y + 1/rho * dp_x - (mu/rho) * (du_xx+ du_yy)
      pde_v    = Y[:,0:1]*dv_x +  Y[:,1:2]*dv_y + 1/rho * dp_y - (mu/rho) * (dv_xx+ dv_yy)
      pde_cont = du_x + dv_y
    
    return [pde_u,pde_v,pde_cont]

And I want to traing "mu" and "rho" in my code. The goal is to achieve optimized values for these coefficients to use in RANS simulations and finally get results with accuracy near the DNS data.

I searched for that and found these commands: 

    mu= tf.Variable( 1, dtype=tf.float32, name="beta")  //placed befor PDEs in code, value 1 is default
    def pde(X, Y): ...

then rest of the steps:

    data = dde.data.PDE(geom,pde,...)
    net = dde.maps.FNN([2] + [64]*5 + [3] , "tanh", "Glorot uniform")
    model = dde.Model(data, net)
    model.compile("adam",lr=1e-3)
    losshistory, train_state = model.train(epochs = 10000)
    dde.optimizers.config.set_LBFGS_options(maxiter= 3000)
    model.compile("L-BFGS")
    losshistory, train_state = model.train()
    dde.saveplot(losshistory, train_state, issave=False, isplot=False)

and:

    optimized_mu = mu.numpy()   //placed after "L-BFGS" optimization
    print(f"alpha_mu: {optimized_mu}")

I used this and there were no error, but no optimization has been occured. There are some sicentific papers which were successful in this and the  optimization has been implemented. They mentioned that used Automatic Differentiation for this purpose and a feed-forward PINN. I made the feed-forward PINN but I dont have any idea for this problem.

Here is the text of one the papers with similar operation:

_The main goal of this study is to implement PINNs algorithm to identify governing equations of a fluid flow for the standard k−ω turbulence model, which includes a continuity equation, two momentum equations, turbulence kinetic energy equation, and the specific dissipation rate equation. **This approach solves an inverse problem by optimizing the five coefficients in k and ω equations.** The values of these coefficients are usually estimated by fitting to experimental data of simple flows. In this method, the prior knowledge in transport equations is imposed on the neural network through physical loss functions. Consequently, the aim is to optimize the coefficients in the k and ω equations by training the model considering limited amount of data obtained from the DNS data of the desired flow. The information related to the geometry and characteristics of the selected flow will be explained in detail in the following section. A schematic of the PINNs algorithm can be found in figure 1. As can be recognized from this figure, **there is a feed-forward neural network in this algorithm** which includes x and y coordinates as input and x-velocity, y-velocity, pressure, turbulent kinetic energy and the specific dissipation rate as outputs._

**Comment by kyouma** [comment]
Have you added the trainable variables list as the `external_trainable_variables` parameter of the `model.compile()` method?

**Comment by LordPontiag** [comment]
> Have you added the trainable variables list as the `external_trainable_variables` parameter of the `model.compile()` method?

Thank you for your reply. At that time when I asked this question no, but finally yes I've done it, and it was due to the problem that you mentioned. I didn't know that I have to set it as a `external_trainable_variables` in the solver and optimizer. Also I found out more about this, I attach the code here for others with same problem:

    def get_trainable_parameter(*variables):
        return [var for var in variables if var.trainable]
    def get_trainable_status(*variables):
        return [var.trainable for var in variables]

    mu  = tf.Variable(1.0,    dtype=tf.float32, trainable = False,  name="alpha_star"   )
    exteranl_VAR = get_trainable_parameters(mu)

    def pde(X, Y): ...
    return ...

    model = dde.Model(data, net)
    model.compile("adam",lr=1e-4, external_trainable_variables = exteranl_VAR)

    optimized_mu_Adam = mu.numpy()

    data_opt = {
    "Coefficient": ["mu"],
    
    "Original Value": [1.0],

    "Optimized Values Adam": [mu],

    "Used for Optimization": get_trainable_status(mu)
               }
    optimization_DF = pd.DataFrame(data_opt)
    print(optimization_DF)

**Comment by gsof1** [comment]
Hello @LordPontiag, i wanted to ask if the problem that you are solving is a turbulent channel flow? If so, did you manage to have good accuracy compared to DNS data?

**Comment by LordPontiag** [comment]
> Hello [@LordPontiag](https://github.com/LordPontiag), i wanted to ask if the problem that you are solving is a turbulent channel flow? If so, did you manage to have good accuracy compared to DNS data?

Hi dear George, No it is not a channel flow, it is more complecated than a channel but it is similar to that in boundary conditions. In fact it is somthing like the backward step and include flow separation. However in case of boundary condition it has an inlet, an outlet, and walls. Regarding the accuacy, In my last simulation I could capture velocity magnitude with MSE of 3.1% and pressure with MSE of 8.85%, Kinetic Energy with MSE of 12%. The pattern and inherent of physics are captured, these errors are more related to differences in values. For instance the component of x velocity in DNS ranges from -0.2m/s to 1m/s, but in my results is in range of 0 to 1m/s. I belive these errors are not unexpected and occur due to physical constrains implementions.
